### PR TITLE
Fix/property drawers

### PR DESCRIPTION
**Pull request by thestonefox**

### CODE REVIEW

File: Editor/Data/Attribute/RestrictedAttributeDrawer.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    using UnityEngine;
+    using UnityEditor;
+
+    [CustomPropertyDrawer(typeof(RestrictedAttribute))]
+    public class RestrictedAttributeDrawer : PropertyDrawer
+    {
+        public override void OnGUI(Rect position, SerializedProperty property, GUIContent label)
+        {
+            EditorGUI.BeginProperty(position, label, property);
+            RestrictedAttribute attrib = attribute as RestrictedAttribute;
```
Comment (bddckr) [inline Editor/Data/Attribute/RestrictedAttributeDrawer.cs]:
```suggestion
            RestrictedAttribute attrib = (RestrictedAttribute)attribute;
```
I don't like the short variable name but can't come up with something else than `restrictedAttribute` which is a bit long.

File: Editor/Data/Attribute/RestrictedAttributeDrawer.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+            bool playingAndActiveAndEnabled = Application.isPlaying && property.serializedObject.targetObject is Behaviour behaviour && behaviour.isActiveAndEnabled;
```
Comment (bddckr) [inline Editor/Data/Attribute/RestrictedAttributeDrawer.cs]:
```suggestion
            bool isPlayingAndActiveAndEnabled = Application.isPlaying && property.serializedObject.targetObject is Behaviour behaviour && behaviour.isActiveAndEnabled;
```
Regular C# naming convention for boolean types.

File: Editor/Data/Attribute/RestrictedAttributeDrawer.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+            bool originalGUIEnabledState = GUI.enabled;
+            if (makeReadOnly)
+            {
+                GUI.enabled = false;
+            }
+
+            Color controlColor = Color.grey;
+
+            Color currentGUIColor = GUI.color;
```
Comment (bddckr) [inline Editor/Data/Attribute/RestrictedAttributeDrawer.cs]:
```suggestion
            Color currentGuiColor = GUI.color;
```

File: Editor/Data/Attribute/RestrictedAttributeDrawer.cs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+            bool muteProperty = (attrib.restrictions & RestrictedAttribute.Restrictions.Muted) != 0;
+
+            bool originalGUIEnabledState = GUI.enabled;
+            if (makeReadOnly)
+            {
+                GUI.enabled = false;
+            }
+
+            Color controlColor = Color.grey;
```
Comment (bddckr) [inline Editor/Data/Attribute/RestrictedAttributeDrawer.cs]:
```suggestion
            const Color controlColor = Color.grey;
```
I think that works.

File: Editor/Data/Attribute/RestrictedAttributeDrawer.cs (hunk 3 above)
Comment (bddckr) [inline Editor/Data/Attribute/RestrictedAttributeDrawer.cs]: I suggest to rename all of these `current*` variables to `original*` like the other one above.

File: Editor/Data/Attribute/RestrictedAttributeDrawer.cs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+                GUI.color = currentGUIColor;
+                EditorStyles.label.normal.textColor = currentNormalTextColor;
+                EditorStyles.label.focused.textColor = currentFocusedTextColor;
+                EditorStyles.label.fontStyle = currentFontStyle;
+            }
+
+            if (makeReadOnly)
+            {
+                GUI.enabled = originalGUIEnabledState;
+            }
```
Comment (bddckr) [inline Editor/Data/Attribute/RestrictedAttributeDrawer.cs]:
```suggestion
            }

```
I feel like we always split control statements with the following expressions with some space like this, but your call!

File: Runtime/Data/Attribute/RestrictedAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
+﻿namespace Zinnia.Data.Attribute
+{
+    using UnityEngine;
+    using System;
+
+    /// <summary>
+    /// Defines the <c>[Restricted]</c> attribute.
```
Comment (bddckr) [inline Runtime/Data/Attribute/RestrictedAttribute.cs]:
```suggestion
    /// Draws the member in a specified restricted way.
```
We should actually fix this on all attributes...

File: Runtime/Data/Attribute/RestrictedAttribute.cs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    /// </summary>
+    public class RestrictedAttribute : PropertyAttribute
+    {
+        [Flags]
+        public enum Restrictions
```
Comment (bddckr) [inline Runtime/Data/Attribute/RestrictedAttribute.cs]: Anything in this class needs docs: Enum, enum's variants, field, constructor. The constructor's summary should be a copy of the type's summary.

File: Editor/Data/Attribute/RestrictedAttributeDrawer.cs (hunk 4 above)
Comment (thestonefox) [reply Editor/Data/Attribute/RestrictedAttributeDrawer.cs]: Color can't be const

Comment (bddckr) [reply Editor/Data/Attribute/RestrictedAttributeDrawer.cs]: 👍

File: Runtime/Data/Attribute/RestrictedAttribute.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    public class RestrictedAttribute : PropertyAttribute
+    {
+        /// <summary>
+        /// The restriction options that can be applied.
+        /// </summary>
+        [Flags]
+        public enum Restrictions
+        {
+            /// <summary>
+            /// The property is always set to ReadOnly
```
Comment (bddckr) [inline Runtime/Data/Attribute/RestrictedAttribute.cs]:
```suggestion
            /// The property is always read-only.
```

File: Runtime/Data/Attribute/RestrictedAttribute.cs [hunk 9]
```diff
@@ -0,0 +1,45 @@
+﻿namespace Zinnia.Data.Attribute
+{
+    using UnityEngine;
+    using System;
+
+    /// <summary>
+    /// Draws the member in a specified restricted way.
```
Comment (bddckr) [inline Runtime/Data/Attribute/RestrictedAttribute.cs]:
```suggestion
    /// Draws the property in a specified restricted way.
```
You used that in the docs you added and I think that's what Unity uses, too.

File: Runtime/Data/Attribute/RestrictedAttribute.cs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+        /// </summary>
+        [Flags]
+        public enum Restrictions
+        {
+            /// <summary>
+            /// The property is always set to ReadOnly
+            /// </summary>
+            ReadOnlyAlways = 1 << 0,
+            /// <summary>
+            /// The property is only set to ReadOnly when the Application is playing and the component is enabled.
```
Comment (bddckr) [inline Runtime/Data/Attribute/RestrictedAttribute.cs]:
```suggestion
            /// The property is read-only when the application is playing and the component is enabled.
```

File: Runtime/Data/Attribute/RestrictedAttribute.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+            /// </summary>
+            ReadOnlyWhenPlayingAndEnabled = 1 << 1,
+            /// <summary>
+            /// The property style is de-emphasized.
```
Comment (bddckr) [inline Runtime/Data/Attribute/RestrictedAttribute.cs]:
```suggestion
            /// The property is de-emphasized.
```

File: Runtime/Data/Attribute/RestrictedAttribute.cs [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+            /// </summary>
+            Muted = 1 << 2
+        }
+
+        /// <summary>
+        /// The restriction options that will be applied.
```
Comment (bddckr) [inline Runtime/Data/Attribute/RestrictedAttribute.cs]:
```suggestion
        /// The restriction options to apply.
```

File: Runtime/Data/Attribute/RestrictedAttribute.cs [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+        /// </summary>
+        public readonly Restrictions restrictions;
+
+        /// <summary>
+        /// Draws the member in a specified restricted way.
```
Comment (bddckr) [inline Runtime/Data/Attribute/RestrictedAttribute.cs]:
```suggestion
        /// Draws the property in a specified restricted way.
```

**Pull request merged**